processor/spanmetricsprocessor/processor.go

```diff
@@ -227,9 +243,17 @@ func (p *processorImp) Capabilities() consumer.Capabilities {
 // It aggregates the trace data to generate metrics, forwarding these metrics to the discovered metrics exporter.
 // The original input trace data will be forwarded to the next consumer, unmodified.
 func (p *processorImp) ConsumeTraces(ctx context.Context, traces pdata.Traces) error {
+	p.lock.Lock()
+	// use defer to pass the output to downstream components as quick as possible.
```

Comment:
By using `defer...` here, the ` p.nextConsumer.ConsumeTraces(ctx, traces)` from https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/7075/files/cb5c3ac2c12f183fbeaa296238298fd493bac957#diff-78f9e7446e0e4fb474786b4b2c262486887bf27479f1f8bc598376d6c3c29b72L245 will be executed before we reset and unlock the processor.

We can't do `err := p.nextConsumer.ConsumeTraces(ctx, traces)` at the very beginning as we need to make sure there is no error in the current processor before we pass the traces to the downstream processors.

Reply:
Any given instance of a processor may be invoked concurrently by any number of receivers or by a receiver that handles incoming requests concurrently.  For instance, an HTTP-based receiver that handles each request in a goroutine may concurrently invoke `next.ConsumeTraces()` on this (or any) processor.

Reply:
Another approach that might be considered is to lean in to concurrency.  Instead of emitting metrics for every invocation of `ConsumeTraces()`, perhaps start a goroutine with a ticker that will periodically take the lock, build the metrics, emit them, and reset the exemplars.  I think the whole of `ConsumeTraces()` would probably still need to be under lock, but it would do less work with each invocation and metrics emission would be more regular.

Reply:
Ah, I see what you folks means now. The reset and unlock should probably happen before the call to the next consumer happens then, otherwise you are blocking more than you need to.

Reply:
@MovieStoreGuy I'm not sure how `atomic` or a channel can do a fast exit as we need to execute this function in synchronization?

@Aneurysm9 How about we have multiple sub-processors to process each single `traces` concurrently and independently. In that way, the lock can be moved to each sub-processor and the upstream can call ConsumeTraces() in a concurrent way. Should we do that in another separate PR or in this one?

Reply:
I'll remove the `defer()` so that we can release the lock earlier.

Reply:
That could be one way of implementing it.  It seemed that there were three steps that the processor performed:

1. aggregate raw data from traces
2. build metrics from the aggregated data and send them to the next metrics consumer
3. reset the storage for aggregated data

I'd suggest having 1) performed synchronously in `ConsumeTraces` with a lock held and then 2) and 3) performed together periodically by a goroutine with a ticker, also holding the lock to ensure that no changes are made to the aggregated data while it is being processed.

Reply:
@Aneurysm9 for step 2 and step 3, as we still need to hold the read lock, which means step 1 will still be blocked periodically. Also, step 1 itself is still synchronized. Would we still worry about this?

As the step 2 and step 3 will be async, what's the best way to handle error propagating in `func ConsumeTraces(ctx context.Context, traces pdata.Traces) error `?

Reply:
I think step 2 and 3 would need a write lock, not just a read lock, as step 3 would reset the storage and wouldn't want to try to swap a read lock for a write lock in between steps 2 and 3.  I don't think it should be an issue that step 1 will be periodically blocked while the metrics are produced.

The traces would be passed on to the next consumer in the pipeline synchronously, after step 1, so that error could be returned directly.  The interesting bit is what to do with any error that comes up while producing metrics asynchronously.  I'm not sure the best way to handle that right now.  It might be worth looking at the batch processor, which also does asynchronous production to the rest of the pipeline.

Reply:
Yes, I have raised similar confusions on related PRs before：https://github.com/open-telemetry/opentelemetry-collector-contrib/pull/6503#discussion_r777992047

And we are currently using an independent goroutine ticker to periodically buildMetrics and consumeMetrics